go/vt/zktopo/serving_graph.go

```diff
@@ -220,14 +238,78 @@ func (zkts *Server) updateTabletEndpoint(oldValue string, oldStat zk.Stat, addr
 	return jscfg.ToJson(addrs), nil
 }
 
+// UpdateTabletEndpoint is part of the topo.Server interface
 func (zkts *Server) UpdateTabletEndpoint(cell, keyspace, shard string, tabletType topo.TabletType, addr *topo.EndPoint) error {
 	path := zkPathForVtName(cell, keyspace, shard, tabletType)
 	f := func(oldValue string, oldStat zk.Stat) (string, error) {
 		return zkts.updateTabletEndpoint(oldValue, oldStat, addr)
 	}
 	err := zkts.zconn.RetryChange(path, 0, zookeeper.WorldACL(zookeeper.PERM_ALL), f)
-	if err == skipUpdateErr || zookeeper.IsError(err, zookeeper.ZNONODE) {
+	if err == errSkipUpdate || zookeeper.IsError(err, zookeeper.ZNONODE) {
 		err = nil
 	}
 	return err
 }
+
+// WatchEndPoints is part of the topo.Server interface
+func (zkts *Server) WatchEndPoints(cell, keyspace, shard string, tabletType topo.TabletType) (<-chan *topo.EndPoints, chan<- struct{}, error) {
+	path := zkPathForVtName(cell, keyspace, shard, tabletType)
+
+	notifications := make(chan *topo.EndPoints, 10)
+	stopWatching := make(chan struct{})
+
+	go func() {
+		for {
+			// set the watch
+			data, _, watch, err := zkts.zconn.GetW(path)
+			if err != nil {
+				if zookeeper.IsError(err, zookeeper.ZNONODE) {
+					// the parent directory doesn't exist
+					notifications <- nil
+				}
+
+				log.Errorf("Cannot set watch on %v, waiting for %s to retry: %v", path, WatchSleepDuration, err)
+				time.Sleep(WatchSleepDuration)
+				continue
+			}
+
+			// get the initial value, send it, or send nil if no
+			// data
+			var ep *topo.EndPoints
+			sendIt := true
+			if len(data) > 0 {
+				ep = &topo.EndPoints{}
+				if err := json.Unmarshal([]byte(data), ep); err != nil {
+					log.Errorf("EndPoints unmarshal failed: %v %v", data, err)
+					sendIt = false
+				}
+			}
+			if sendIt {
+				notifications <- ep
+			}
+
+			// now act on the watch
+			select {
+			case event, ok := <-watch:
```

Comment:
Should send "event" to "notifications" if 'ok'?
I also think we need a for loop around "select" to continuously process events.